### PR TITLE
Migrating nonce type from number to string

### PR DESCRIPTION
**Pull request by hcho112**
## Motivation
#220 

## Description
This PR contains implementations to migrate number typed nonce to string.
Added a string type check to throw error if `nonce` with `number` is given.
Updated all existing tests

## Checklist
- [x] Read the [contributing](https://github.com/near/near-api-js/blob/master/CONTRIBUTING.md) guidelines
- [x] Commit messages follow the [conventional commits](https://www.conventionalcommits.org/) spec
- [x] Performed a self-review of the PR
- [ ] Added automated tests
- [x] Manually tested the change

### CODE REVIEW

**Review by esaminu — APPROVED** [review]
LGTM 🚀  

Can we use conventional commit syntax on the PR description and mark this as a breaking change?

**Comment by hcho112** [comment]
wait sorry lets not merge this yet. I had deep dive discussion with @MaximusHaximus and it is now essential to make this on this PR.

**Comment by hcho112** [comment]
After my discovery on the fact that RPC is returning `nonce` on `accessKey` as number, I have re-write this PR to cater current implementation as well as future proof. (by wrapping `nonce` with `BN` so that it is guaranteed to operate consistently when it's a number or string is given)

File: packages/near-api-js/src/account.ts [hunk 1]
```diff
@@ -579,13 +585,15 @@ export class Account {
...
             finality: 'optimistic'
         });
+        // Replace raw nonce into a new BN
+        const newResponse = { keys: response?.keys?.map((key) => ({ ...key, access_key: { ...key.access_key, nonce: new BN(key.access_key.nonce) } }))};
         // A breaking API change introduced extra information into the
         // response, so it now returns an object with a `keys` field instead
         // of an array: https://github.com/nearprotocol/nearcore/pull/1789
-        if (Array.isArray(response)) {
-            return response;
+        if (Array.isArray(newResponse)) {
```
Comment (morgsmccauley) [inline packages/near-api-js/src/account.ts:593]: This check seems irrelevant now that `newResponse` is always an object. Do we need additional logic here to cater for `view_access_key_list` returning either an array or an object, or can we trust that it will always be an object of the shape `{ keys: [] }`

File: packages/near-api-js/src/account.ts [hunk 2]
```diff
@@ -293,13 +294,18 @@ export class Account {
         }
 
         try {
-            const accessKey = await this.connection.provider.query<AccessKeyView>({
+            const rawAccessKey = await this.connection.provider.query<AccessKeyViewRaw>({
```
Comment (morgsmccauley) [inline packages/near-api-js/src/account.ts:297]: Would it be better to place this coercion logic within `JsonRpcProvider`? `nonce` is only really used in this file, but if we were to use these same `query` `methods` else where, we'd most likely have to duplicate this logic. Similarly, someone adding a new `query` type in `JsonRpcProvider` which contains `nonce` may not know that coercion is necessary because it happens within this file.

We'd have to inspect the incoming `request_type` which doesn't seem ideal, but I think the centralization of it would be beneficial.

Thoughts?

**Review by morgsmccauley — CHANGES_REQUESTED** [review]
Mostly looks good, just a couple questions :)

File: packages/near-api-js/src/account.ts (hunk 2 above)
Comment (hcho112) [reply packages/near-api-js/src/account.ts:297]: First of all, changes on this PR is based on the mindset of `nonce` should be returned as a `string` NOT a `number`. (It is currently returning as a number) The problem with returning with number is that, javascript `number` type is less than what max `u64` can be as an integer. So at some point, it will be essential for RPC to make this change for the sake of precision on `nonce`. 

So, changes on this PR allows to support both `number` and `string` type of nonce where it will become a future proof regardless when RPC is making the `nonce` type change. So assuming that `nonce` is returned as `string` that is larger than js `number` integer, in order to increment, we can't just use `++`. So it should still need to be wrapped with `new BN`, wherefore this logic is actually unavoidable.

Comment (morgsmccauley) [reply packages/near-api-js/src/account.ts:297]: What I am suggesting here is we move this logic in to `JsonRpcProvider` so that it is encapsulated and will be reused if these queries are ever used elsewhere.

But on second thought it may be better to leave this as is, and deal with that case if/when we need.

**Review by morgsmccauley — APPROVED** [review]
LGTM
